src/ring_discord_auth/validation.clj

```diff
@@ -0,0 +1,181 @@
+                      (do
+                        (.write bos next)
+                        (recur (.read ^InputStream input))))))
+    ByteBuffer (let [len (.remaining ^ByteBuffer input)
+                     result (byte-array len)]
+                 (.get ^ByteBuffer input result)
+                 result)))
+
+(defn encode
+  "Encodes the given string to a byte array using the given charset/encoding.
+
+  Returns `nil` if the charset is not available or if it doesn't support encoding."
+  {:test (examples encode Arrays/equals
+                   ["Hello, world!", "utf8" #_=> (byte-array [0x48 0x65 0x6c 0x6c 0x6f 0x2c 0x20 0x77 0x6f 0x72 0x6c 0x64 0x21])]
+                   ["Another one", "unknown?" #_=> nil])}
+  ^bytes [^String str ^String charset-name]
+  (if-let-all [^Charset cs (try (Charset/forName charset-name) (catch UnsupportedCharsetException _ nil) (catch IllegalCharsetNameException _ nil))
+               ^CharsetEncoder encoder (try (.newEncoder cs) (catch UnsupportedOperationException _ nil))]
+              (when (.canEncode encoder str)
+                (read-all-bytes (.encode encoder (CharBuffer/wrap str))))))
+
+(defn hex->bytes
+  "Converts the given string representing a hexadecimal number to a byte array.
+  Each byte in the resulting array comes from 2 digits in the string.
+
+  If the string cannot be converted, returns `nil`"
+  ^bytes [^String hex-str]
+  (let [len (count hex-str)
+        result (byte-array (quot (inc len) 2))]
+    (try
+      (doseq [[i hex-part] (map-indexed vector (map (partial apply str) (partition-all 2 hex-str)))]
+        (aset result i (unchecked-byte (Short/parseShort hex-part 16))))
+      result
+      (catch NumberFormatException _ nil))))
+
+(defn bytes->hex
+  "Convert byte array to hex string."
+  [^bytes byte-array]
+  (let [hex [\0 \1 \2 \3 \4 \5 \6 \7 \8 \9 \a \b \c \d \e \f]]
+    (letfn [(hexify-byte [b]
+              (let [v (bit-and b 0xFF)]
+                [(hex (bit-shift-right v 4)) (hex (bit-and v 0x0F))]))]
+      (str/join (mapcat hexify-byte byte-array)))))
+
+(defn generate-keypair
+  "Generate Ed25519 key pair.
+
+  Return {:private `Ed25519PrivateKeyParameters`
+          :public `Ed25519PublicKeyParameters`}"
+  []
+  (let [random (SecureRandom.)
+        kpg    (Ed25519KeyPairGenerator.)
+        _ (.init kpg (Ed25519KeyGenerationParameters. random))
+        key-pair (.generateKeyPair kpg)]
+    {:private (cast Ed25519PrivateKeyParameters (.getPrivate key-pair))
+     :public  (cast Ed25519PublicKeyParameters (.getPublic key-pair))}))
+
+(defn new-signer
+  "Return new instance of `Ed25519Signer` initialized by private key"
+  [private-key]
+  (let [signer (Ed25519Signer.)]
+    (.init signer true private-key)
+    signer))
+
+(defn sign
+  "Generate signature for msg byte array.
+
+  Return byte array with signature."
+  [^Ed25519Signer signer msg-bytes]
+  (.update signer msg-bytes 0 (alength msg-bytes))
+  (.generateSignature signer))
```

Comment:
I feel like these functions should be part of the test namespace since they are not needed by users of this library. Feel free to correct me on that

Reply:
:+1: